File: C/17299.c
#include <stdio.h>  // 오등큰수
#include <stdlib.h>

int cnt[1000001] = { 0, };
int input[1000001];

typedef struct
{
	int* stk;
	int ptr;
}STACK;

int is_empty(STACK* s)
{
	if (s->ptr <= -1)
		return 1;
	else
		return 0;
}

void push(STACK* s, int data)
{
	s->stk[++s->ptr] = data;
}

int pop(STACK* s)
{
	return s->stk[s->ptr--];
}

int top(STACK* s)
{
	return s->stk[s->ptr];
}


int main()
{
	int n;
	scanf("%d", &n);

	for (int i = 0; i < n; i++)
	{
		scanf("%d", &input[i]);
		cnt[input[i]]++;
	}

	STACK s;
	s.stk = malloc(n * sizeof(int)); s.ptr = -1;

	for (int i = 0; i < n; i++)
	{
		while (!is_empty(&s) && cnt[input[top(&s)]] < cnt[input[i]])
		{
			input[top(&s)] = input[i];
			pop(&s);
		}
		push(&s, i);
	}

	while (!is_empty(&s))
	{
		input[top(&s)] = -1;
		pop(&s);
	}

	for (int i = 0; i < n; i++)
		printf("%d ", input[i]);

	return 0;
}

File: JAVA/src/_17299.java
import java.io.*;
import java.util.Stack;

// 오등큰수
public class _17299 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int n = Integer.parseInt(br.readLine());
        String[] input = br.readLine().split(" ");

        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(input[i]);
        }

        // cnt 배열에 숫자가 나온 횟수를 count한 값 저장
        int[] cnt = new int[1000001];
        for (int i = 0; i < n; i++) {
            cnt[arr[i]]++;
        }

        int[] ans = new int[n];
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && (cnt[arr[stack.peek()]] < cnt[arr[i]])) {
                ans[stack.pop()] = arr[i];
            }

            stack.push(i);
        }

        while (!stack.isEmpty()) {
            ans[stack.pop()] = -1;
        }

        for (int num : ans) {
            bw.write(num + " ");
        }

        bw.flush();
        bw.close();
    }
}
